lib/internal/webstreams/adapters.js

```diff
@@ -0,0 +1,909 @@
+
+const {
+  WriteWrap,
+  ShutdownWrap,
+  kReadBytesOrError,
+  kLastWriteWasAsync,
+  streamBaseState,
+} = internalBinding('stream_wrap');
+
+const finished = require('internal/streams/end-of-stream');
+
+const { UV_EOF } = internalBinding('uv');
+
+/**
+ * @typedef {import('../../stream').Writable} Writable
+ * @typedef {import('../../stream').Readable} Readable
+ * @typedef {import('./writablestream').WritableStream} WritableStream
+ * @typedef {import('./readablestream').ReadableStream} ReadableStream
+ *
+ * @typedef {import('../abort_controller').AbortSignal} AbortSignal
+ */
+
+/**
+ * @param {Writable} streamWritable
+ * @returns {WritableStream}
+ */
+function newWritableStreamFromStreamWritable(streamWritable) {
+  if (typeof streamWritable?._writableState !== 'object') {
+    throw new ERR_INVALID_ARG_TYPE(
+      'streamWritable',
+      'stream.Writable',
+      streamWritable);
+  }
+
+  if (streamWritable.destroyed || !streamWritable.writable) {
+    const writable = new WritableStream();
+    writable.close();
+    return writable;
+  }
+
+  const highWaterMark = streamWritable.writableHighWaterMark;
+  const strategy =
+    streamWritable.writableObjectMode ?
+      new CountQueuingStrategy({ highWaterMark }) :
+      { highWaterMark };
+
+  let controller;
+  let backpressurePromise;
+  let closed;
+  let errored = false;
+
+  function onDrain() {
+    if (backpressurePromise !== undefined)
+      backpressurePromise.resolve();
+  }
+
+  const cleanup = finished(streamWritable, (error) => {
+    cleanup();
+    // This is a protection against non-standard, legacy streams
+    // that happen to emit an error event again after finished is called.
+    streamWritable.on('error', () => {});
+    if (error != null) {
+      errored = true;
+      if (backpressurePromise !== undefined)
+        backpressurePromise.reject(error);
+      // If closed is not undefined, the error is happening
+      // after the WritableStream close has already started.
+      // We need to reject it here.
+      if (closed !== undefined) {
+        closed.reject(error);
+        closed = undefined;
+      }
+      controller.error(error);
+      controller = undefined;
+      return;
+    }
+
+    if (closed !== undefined) {
+      closed.resolve();
+      closed = undefined;
+      return;
+    }
+    controller.error(new ERR_STREAM_PREMATURE_CLOSE());
+    controller = undefined;
+  });
+
+  streamWritable.on('drain', onDrain);
+
+  return new WritableStream({
+    start(c) { controller = c; },
+
+    async write(chunk) {
+      if (streamWritable.writableNeedDrain || !streamWritable.write(chunk)) {
+        backpressurePromise = createDeferredPromise();
+        return PromisePrototypeFinally(
+          backpressurePromise.promise, () => {
+            backpressurePromise = undefined;
+          });
+      }
+    },
+
+    abort(reason) {
+      if (!streamWritable.destroyed && !errored)
+        destroy(streamWritable, reason);
```

Comment:
```suggestion
      stream.destroy(streamWritable, reason);
```

I don't think you need the condition.